### PR TITLE
BackendService naming for NEG backend services & healthchecks

### PR DESCRIPTION
**Pull request by nicksardo**
Continuation of https://github.com/kubernetes/ingress-gce/pull/239

More changes:
 - Neg suffix hash uses the UID truncated to 8 chars.
 - Neg suffix hash input fields are separated by a delimiter
 - ServicePort now has an extension function to return the correct name. 
 - `namer.Backend` is renamed to `namer.IGBackend` to prevent confusion between IG/NEG.

TODO in another PR:
Rewrite unit tests to not care what kind of serviceport their given.

### CODE REVIEW

File: pkg/loadbalancers/l7.go [hunk 1]
```diff
@@ -137,7 +137,7 @@ func (l *L7) checkUrlMap() (err error) {
 	if l.runtimeInfo.UrlMap == nil {
 		return fmt.Errorf("cannot create urlmap without internal representation")
 	}
-	defaultBackendName := l.namer.Backend(l.runtimeInfo.UrlMap.DefaultBackend.NodePort)
+	defaultBackendName := l.namer.IGBackend(l.runtimeInfo.UrlMap.DefaultBackend.NodePort)
```
Comment (rramkumar1) [inline pkg/loadbalancers/l7.go]: How come you bypass ServicePort.BackendName() here and just call IGBackend() directly?

Comment (rramkumar1) [inline pkg/loadbalancers/l7.go]: How come you use IGBackend() directly here instead of ServicePort.BackendName()?

**Comment by nicksardo** [comment]
Manual test looked good. Nice job @agau4779

**Comment by rramkumar1** [comment]
Looks fine to me. Will leave @freehan  to give LGTM.

File: pkg/utils/namer.go [hunk 2]
```diff
@@ -239,13 +247,13 @@ func (n *Namer) NameBelongsToCluster(name string) bool {
 	return components.ClusterName == clusterName || n.negBelongsToCluster(name)
 }
 
-// BeName constructs the name for a backend.
-func (n *Namer) Backend(port int64) string {
+// IGBackend constructs the name for a backend.
```
Comment (freehan) [inline pkg/utils/namer.go]: comment: IGBackend constructs the name for a backend service targeting instance groups.

File: pkg/utils/namer.go [hunk 3]
```diff
@@ -216,6 +223,19 @@ func (n *Namer) ParseName(name string) *NameComponents {
 	}
 }
 
+// negBelongsToCluster checks that the uid is present and a substring of the
+// cluster uid, since the NEG naming schema truncates it to 8 characters.
+// This is only valid for NEGs and Backends on NEG.
+func (n *Namer) negBelongsToCluster(name string) bool {
```
Comment (freehan) [inline pkg/utils/namer.go]: nit: negNameBelongsToCluster?

comment: This is only valid for NEGs, BackendServices and Healthchecks for NEG

**Comment by nicksardo** [comment]
Thanks for both of your reviews. Fixed comments.

**Pull request merged**